File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        res = l1
        carry = 0
        prev = None
        while(l1 and l2):
            add = l1.val + l2.val + carry
            carry = 0
            if add >= 10:
                carry = add // 10
                add = add % 10
                isCarry = True
            else:
                isCarry = False
            l1.val = add
            if l1.next == None:
                prev = l1
            l1 = l1.next
            l2 = l2.next

        while(l1):
            add = l1.val + carry
            carry = 0
            if add >= 10:
                carry = add // 10
                add = add % 10
                isCarry = True
            else:
                isCarry = False
            l1.val = add
            if l1.next == None:
                prev = l1
            l1 = l1.next
        while(l2):
            add = l2.val + carry
            carry = 0
            if add >= 10:
                carry = add // 10
                add = add % 10
                isCarry = True
            else:
                isCarry = False
            prev.next = ListNode(add)
            prev = prev.next
            l2 = l2.next
        if isCarry:
            prev.next = ListNode(carry)
        return res

File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode top = l1;
        ListNode bottom = l2;
        int carry = 0;
        ListNode head = new ListNode(0,null);
        ListNode temp = head;
        while(top!= null && bottom != null){
            ListNode newNode = new ListNode((top.val + bottom.val + carry) %10, null);
            head.next = newNode;
            head = head.next;
            carry = (top.val + bottom.val + carry)/10;
            top = top.next;
            bottom = bottom.next;
        }
        while(top!= null){
            ListNode newNode = new ListNode((top.val + carry)%10, null);
            head.next = newNode;
            head = head.next;
            carry = (top.val + carry)/10;
            top = top.next;
        }
        while(bottom!= null){
            ListNode newNode = new ListNode((bottom.val + carry)%10, null);
            head.next = newNode;
            head = head.next;
            carry = (bottom.val + carry)/10;
            bottom = bottom.next;
        }
        if(carry > 0){
            ListNode newNode = new ListNode(carry, null);
            head.next = newNode;
        }
        return temp.next;
    }
}